### PR TITLE
Detect diff. between BW items and secrets

### PR DESCRIPTION
**Pull request by hongkailiu**
https://issues.redhat.com/browse/DPTP-645

/cc @openshift/openshift-team-developer-productivity-test-platform

### CODE REVIEW

File: cmd/ci-secret-bootstrap/main.go [hunk 1]
```diff
@@ -212,19 +215,26 @@ func constructSecrets(config []secretConfig, bwClient bitwarden.Client) (map[str
...
+				if !kerrors.IsNotFound(err) {
 					return err
 				}
-				if _, err := secretsGetter.Secrets(secret.Namespace).Update(secret); err != nil {
+				if _, err := secretsGetter.Secrets(secret.Namespace).Create(secret); err != nil {
 					return err
 				}
 			}
+			if !force && !equality.Semantic.DeepEqual(existingSecret.Data, secret.Data) {
+				return fmt.Errorf("found unsynchronized secret: %s:%s/%s", cluster, secret.Namespace, secret.Name)
```
Comment (stevekuznetsov) [inline cmd/ci-secret-bootstrap/main.go]: You should paste the diff here

File: cmd/ci-secret-bootstrap/main.go [hunk 2]
```diff
@@ -212,19 +215,26 @@ func constructSecrets(config []secretConfig, bwClient bitwarden.Client) (map[str
...
-func updateSecrets(secretsGetters map[string]coreclientset.SecretsGetter, secretsMap map[string][]*coreapi.Secret) error {
+func updateSecrets(secretsGetters map[string]coreclientset.SecretsGetter, secretsMap map[string][]*coreapi.Secret, force bool) error {
 	for cluster, secrets := range secretsMap {
 		for _, secret := range secrets {
 			logrus.Infof("handling secret: %s:%s/%s", cluster, secret.Namespace, secret.Name)
 			secretsGetter := secretsGetters[cluster]
-			if _, err := secretsGetter.Secrets(secret.Namespace).Create(secret); err != nil {
-				if !kerrors.IsAlreadyExists(err) {
+			existingSecret, err := secretsGetter.Secrets(secret.Namespace).Get(secret.Name, meta.GetOptions{})
+			if err != nil {
```
Comment (stevekuznetsov) [inline cmd/ci-secret-bootstrap/main.go]: Couldn't this be simplified a bit

```go
if kerrors.IsNotFound(err) {
    if _, err := secretsGetter.Secrets(secret.Namespace).Create(secret); err != nil {
        return err
    }
} else if err != nil {
    return err
}
```

File: cmd/ci-secret-bootstrap/main.go [hunk 3]
```diff
@@ -212,24 +216,40 @@ func constructSecrets(config []secretConfig, bwClient bitwarden.Client) (map[str
...
 				}
+			} else if err != nil {
+				return err
+			}
+			if !force && !equality.Semantic.DeepEqual(secret.Data, existingSecret.Data) {
+				logrus.Errorf("actual %s:%s/%s differs the expected:\n%s", cluster, secret.Namespace, secret.Name,
+					cmp.Diff(bytesMapToStringMap(secret.Data), bytesMapToStringMap(existingSecret.Data)))
+				return fmt.Errorf("found unsynchronized secret: %s:%s/%s", cluster, secret.Namespace, secret.Name)
+			}
+			if _, err := secretsGetter.Secrets(secret.Namespace).Update(secret); err != nil {
```
Comment (droslean) [inline cmd/ci-secret-bootstrap/main.go]: If we hit `kerrors.IsNotFound(err)` you create the ticket above and there is no reason to `Update` it again.

File: cmd/ci-secret-bootstrap/main.go [hunk 4]
```diff
@@ -212,24 +216,40 @@ func constructSecrets(config []secretConfig, bwClient bitwarden.Client) (map[str
...
-				if _, err := secretsGetter.Secrets(secret.Namespace).Update(secret); err != nil {
+			existingSecret, err := secretsGetter.Secrets(secret.Namespace).Get(secret.Name, meta.GetOptions{})
+			if kerrors.IsNotFound(err) {
+				if _, err := secretsGetter.Secrets(secret.Namespace).Create(secret); err != nil {
 					return err
 				}
+			} else if err != nil {
+				return err
+			}
+			if !force && !equality.Semantic.DeepEqual(secret.Data, existingSecret.Data) {
```
Comment (droslean) [inline cmd/ci-secret-bootstrap/main.go]: Lets do that right after we Get the secret.
 And we want to update only if that passes right?

It seems that you need something like:

```go
			if existingSecret, err := secretsGetter.Secrets(secret.Namespace).Get(secret.Name, meta.GetOptions{}); err == nil {
				if !force && !equality.Semantic.DeepEqual(secret.Data, existingSecret.Data) {
					// logrus....
					return fmt.Errorf("found unsynchronized secret: %s:%s/%s", cluster, secret.Namespace, secret.Name)
				}

				if _, err := secretsGetter.Secrets(secret.Namespace).Update(secret); err != nil {
					return err
				}

			} else if kerrors.IsNotFound(err) {
				if _, err := secretsGetter.Secrets(secret.Namespace).Create(secret); err != nil {
					return err
				}
			} else
				return err
			}
```

File: cmd/ci-secret-bootstrap/main.go (hunk 1 above)
Comment (hongkailiu) [reply cmd/ci-secret-bootstrap/main.go]: Done.

File: cmd/ci-secret-bootstrap/main.go (hunk 2 above)
Comment (hongkailiu) [reply cmd/ci-secret-bootstrap/main.go]: Updated with more fix from Nikos.

File: cmd/ci-secret-bootstrap/main.go (hunk 3 above)
Comment (hongkailiu) [reply cmd/ci-secret-bootstrap/main.go]: Done.

File: cmd/ci-secret-bootstrap/main.go (hunk 4 above)
Comment (hongkailiu) [reply cmd/ci-secret-bootstrap/main.go]: Good catch. Updated.

**Pull request merged**